akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -457,10 +459,18 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
    */
   final protected def tryPull[T](in: Inlet[T]): Unit = if (!isClosed(in)) pull(in)
 
+  /**
+   * Requests to stop receiving events from a given input port. Cancelling clears any ungrabbed elements from the port.
+   *
+   * It is recommended to provide a cause.
+   */
+  @deprecated("Use other overload to provide a cause for cancellation or use SubscriptionWithCancelException.NoCause", since = "2.6.0")
+  final protected def cancel[T](in: Inlet[T]): Unit = cancel(in, SubscriptionWithCancelException.NoCause)
+
   /**
    * Requests to stop receiving events from a given input port. Cancelling clears any ungrabbed elements from the port.
    */
-  final protected def cancel[T](in: Inlet[T]): Unit = interpreter.cancel(conn(in))
+  final protected def cancel[T](in: Inlet[T], cause: Throwable): Unit = interpreter.cancel(conn(in), cause)
```

Comment:
What about we actually model it with a type instead of pretend-modeling it with a Throwable constant as no-value?

Something like `Try[Done]` or a custom `sealed trait Cancellation` that can be `Normal("I'm take and I took all I wanted")` or `Failure(error)`?